### PR TITLE
[shared_preferences] Migrate examples to NNBD

### PR DESCRIPTION
**Pull request by stuartmorgan**
Migrate all examples to NNBD so that the example apps will run in strong mode.

Converts macOS example to use the platform interface to remove circular dependencies (code is based on the Linux and Windows versions which already use essentially that approach).

## Pre-launch Checklist

- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [ ] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [ ] I updated CHANGELOG.md to add a description of the change.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning

### CODE REVIEW

File: packages/shared_preferences/shared_preferences_linux/example/lib/main.dart [hunk 1]
```diff
@@ -24,22 +24,22 @@ class MyApp extends StatelessWidget {
...
+  SharedPreferencesDemo({Key? key}) : super(key: key);
 
   @override
   SharedPreferencesDemoState createState() => SharedPreferencesDemoState();
 }
 
 class SharedPreferencesDemoState extends State<SharedPreferencesDemo> {
   final prefs = SharedPreferencesLinux.instance;
-  Future<int> _counter;
+  Future<int>? _counter;
```
Comment (cyanglaz) [inline packages/shared_preferences/shared_preferences_linux/example/lib/main.dart:35]: nits: Maybe this should be `late Future<int> _counter'`?

File: packages/shared_preferences/shared_preferences_windows/example/lib/main.dart [hunk 2]
```diff
@@ -24,22 +24,22 @@ class MyApp extends StatelessWidget {
...
+  SharedPreferencesDemo({Key? key}) : super(key: key);
 
   @override
   SharedPreferencesDemoState createState() => SharedPreferencesDemoState();
 }
 
 class SharedPreferencesDemoState extends State<SharedPreferencesDemo> {
   final prefs = SharedPreferencesWindows.instance;
-  Future<int> _counter;
+  Future<int>? _counter;
```
Comment (cyanglaz) [inline packages/shared_preferences/shared_preferences_windows/example/lib/main.dart:35]: nits: `late Future<int> _counter;`

**Review by cyanglaz — APPROVED** [review]
LGTM % nits

File: packages/shared_preferences/shared_preferences_linux/example/lib/main.dart (hunk 1 above)
Comment (stuartmorgan) [reply packages/shared_preferences/shared_preferences_linux/example/lib/main.dart:35]: Fixed in all examples. I think I was confusing myself by thinking of it as being initialized asynchronously so needing to be `?`, forgetting that the *value* is async, but the future itself isn't.

**Pull request merged**